### PR TITLE
DPE-2217 premptively switchover

### PR DESCRIPTION
**Pull request by paulomach**
Port of [k8s](https://github.com/canonical/mysql-k8s-operator/pull/291)

### CODE REVIEW

**Review by shayancanonical — DISMISSED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -248,6 +249,22 @@ def _on_database_storage_detaching(self, _) -> None:
         if not self._mysql.is_instance_in_cluster(self.unit_label):
             return
 
+        def _get_leader_unit() -> Optional[Unit]:
+            """Get the leader unit."""
+            for unit in self.peers.units:
+                if unit.is_leader():
+                    return unit
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:256]: too good to be true

[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py]: 3004f54

File: src/charm.py [hunk 2]
```diff
@@ -165,6 +167,11 @@ def _on_leader_elected(self, _) -> None:
                 self.set_secret(
                     "app", required_password, generate_random_password(PASSWORD_LENGTH)
                 )
+        self.unit_peer_data.update({"leader": "true"})
+
+    def _on_leader_settings_changed(self, _) -> None:
```
Comment (shayancanonical) [inline src/charm.py:172]: is leader_settings_changed emitted for the leader unit after leader_elected is emitted?

Comment (paulomach) [reply src/charm.py:172]: just for non-leaders, as observed and in docs

**Review by shayancanonical — APPROVED** [review]

File: src/charm.py [hunk 3]
```diff
@@ -248,6 +255,22 @@ def _on_database_storage_detaching(self, _) -> None:
...
+        def _get_leader_unit() -> Optional[Unit]:
+            """Get the leader unit."""
+            for unit in self.peers.units:
+                if self.peers.data[unit]["leader"] == "true":
+                    return unit
+
+        if self._mysql.get_primary_label() == self.unit_label and not self.unit.is_leader():
+            # Preemptively switch primary to unit leader
+            logger.info("Switching primary to the first unit")
+            if leader_unit := _get_leader_unit():
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:267]: The gist is the leader is a preferred unit from the original deployment, not an on-demand scaled up unit that will be removed eventually, so less likely to be removed

[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:267]: not guaranteed

[the remark this answers is not in the document]
Comment (taurus-forever) [reply src/charm.py:267]: Honestly... I would go with `Juju leader == MySQL Primary`.

**Review by taurus-forever — APPROVED** [review]
LGTM, this should work as well.

**Pull request merged**